Tutorial from Manon but cute (#33)
scilus/nf-neuro-documentation: src/content/docs/pipeline/tutorial_steps_1-2.mdx

:::

<RunIcon class="inline-icon text-blue-300" aria-hidden /> **Now, you can run nextflow..**

   <CommandOutputs>
   <span slot="command">
    ```bash
    nextflow run main.nf --input data -profile docker
    ```
   </span>
   <span slot="output">
    ```bash
    [sub-003_ses-01_dir-AP, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bval, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bvec, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.nii.gz]
    ```
   </span>
   </CommandOutputs>

   Each element in the output channel is a tuple containing:

    <html>
        <li>A unique key identifier (subject/session)</li>
        <li>The matching .bval file</li>
        <li>The matching .bvec file</li>
        <li>The matching .nii.gz file (DWI image)</li>
    </html>
  
    And following this format : 

    ```bash
    [ subject_session_id,
      /path/to/subject/session/dwi/*dwi.bval,
      /path/to/subject/session/dwi/*dwi.bvec ,
      /path/to/subject/session/dwi/*dwi.nii.gz]
    ```

### 2. Set correctly the Subject and session ID

> Now let's modify the input structure to make `sub-003_ses-01_dir-AP` become `sub-003_ses-01` 
> using the following structure and `it`. Replace the fromFilePairs call.

Reviewer: ```suggestion
Now let's modify the input structure to make the key identifier `sub-003_ses-01_dir-AP` become `sub-003_ses-01`.
We still use the current structure, but with an additional item mapping using `it`. You can replace the previous `fromFilePairs` call the `after` section.
```

    ```nextflow
        dwi_channel = Channel.fromFilePairs("$input/**/**/dwi/*dwi.{nii.gz,bval,bvec}", size: 3, flat: true)
                 { it.parent.parent.parent.name + "_" + it.parent.parent.name}`;
    ```
  </TabItem>
</Tabs>

<RunIcon class="inline-icon text-blue-300" aria-hidden /> **Now, you can run nextflow..**

   <CommandOutputs>
   <span slot="command">
    ```bash
    nextflow run main.nf --input data -profile docker
    ```
   </span>
   <span slot="output">
    ```bash
    [sub-003_ses-01, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bval, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bvec, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.nii.gz]
    ```
   </span>
   </CommandOutputs>

:::tip[Let's break down these changes]
```
{ it.parent.parent.parent.name + "_" + it.parent.parent.name}
```
    This is a closure that defines how to create the grouping key for the file pairs. 
    It's using the names of the parent directories to create a unique identifier, 
    so you need to add as many "parent" as necessary to fit your data structure.

>     <html>
>     <li>`it` corresponds to the file name (here: /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bval).</li>
>     <li>`.parent` corresponds to the parent folder.</li>
>     </html>

Reviewer: ```suggestion
    <Steps>

    1. `it` corresponds to the file name (here: /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bval).

    2. `.parent` corresponds to the parent folder.

    </Steps>
```
I think using 1/2 would make it clearer that it sequencial, meaning that you use it `it.parent.parent...`

>     To get subjectID `sub-003` from `it`: 

Reviewer: ```suggestion
    **To get subjectID `sub-003` from `it`:**
```
Make it stand out more

     <html>
    <li>`sub-003_ses-01_dir-AP_dwi.bval` `.parent`-> dwi</li>
    <li>`dwi` `.parent`-> ses-01</li>
    <li>`ses-01` `.parent`-> sub-003</li>
    </html>
      

>     it.parent.parent.parent.name -> `sub-003`

Reviewer: ```suggestion
    `it.parent.parent.parent.name` -> `sub-003`
```

>     To get ses-name `ses-01` from `it`: 

Reviewer: ```suggestion
    **To get ses-name `ses-01` from `it`:**
```

    <html>
    <li>`sub-003_ses-01_dir-AP_dwi.bval` `.parent`-> dwi</li>
    <li>`dwi` `.parent`-> ses-01</li>
    </html>

>     it.parent.parent.name -> `ses-01`

Reviewer: ```suggestion
    `it.parent.parent.name` -> `ses-01`
```

    ```
   </TabItem>
  <TabItem label="After">
    ```nextflow
        dwi_channel = Channel.fromFilePairs("$input/**/**/dwi/*dwi.{nii.gz,bval,bvec}", size: 3, flat: true)
            { it.parent.parent.parent.name + "_" + it.parent.parent.name}
            .map{ sid, bvals, bvecs, dwi -> [ [id: sid], dwi, bvals, bvecs ] } // Reordering the inputs.
    ```
  </TabItem>
</Tabs>

<RunIcon class="inline-icon text-blue-300" aria-hidden /> **Now, you can run nextflow..**

   <CommandOutputs>
   <span slot="command">
    ```bash
    nextflow run main.nf --input data -profile docker
    ```
   </span>
   <span slot="output">
    ```bash frame="none"      
    [[id:sub-003_ses-01], /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.nii.gz, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bval, /workspaces/nf-neuro-tutorial_test/data/sub-003/ses-01/dwi/sub-003_ses-01_dir-AP_dwi.bvec]
   ```
   </span>
   </CommandOutputs>

:::tip[Let's break down these changes]
    `.map{ sid, bvals, bvecs, dwi -> [ [id: sid], dwi, bvals, bvecs ] }`

    This is a map operation that reorganizes the emitted data.

>     It takes the sid (sample ID), bvals, bvecs, and dwi files and reorganizes them into a new structure.
> 
>     The output is a channel where each element is a tuple containing:
> 
>         - [id: sid] corresponds to [id:sub-003_ses-01]
>         - dwi corresponds to the dwi.nii.gz file
>         - bvals corresponds to the .bval file
>         - bvecs corresponds to the .bvec file

Reviewer: ```suggestion
    It takes the `sid` (sample ID), `bvals`, `bvecs`, and `dwi` files and reorganizes them into a new structure.

    The output is a channel where each element is a tuple containing:

        - `[id: sid]` corresponds to `[id:sub-003_ses-01]`
        - `dwi` corresponds to the `dwi.nii.gz` file
        - `bvals` corresponds to the `.bval` file
        - `bvecs` corresponds to the `.bvec` file
```
